### ISSUE TITLE
mocking nested dependencies

### ISSUE THREAD

**Issue by sarneeh** [boblauer/mock-require#34]
Consider this test:

subject:
```js
const dep = require('./dep');

module.exports = () => {
    return dep();
}
```

dep:
```js
const uuid = require('uuid/v4');

module.exports = () => {
    return uuid();
}
```

test:
```js
const t = require('tap');
const mock = require('mock-require');

let subject = require('./subject');

t.test('should return uuid', t => {
  mock('uuid/v4', () => '12345');
  subject = mock.reRequire('./subject');

  const result = subject();

  t.equal(result, '12345');
  t.end();
});
```

This test fails because the nested `uuid/v4` dependency is not actually mocked. Why is that and how can I make it work?

**Comment by sarneeh** [comment]
I figured it out that I can fix this test by using `reRequire` on the nested module. But this is very unintuitive. On every case where I'd need to use it I'd need to `reRequire` both the direct test subject and the nested module which uses some dependency which is pretty cumbersome.

To be clear: I'm not doing this for unit tests as this is a clear anti-pattern. I need this for integration testing - I'd like to mock some underlying modules and make them fail to test some few error cases.

Is it the only way to achieve what I want to do?

**Comment by boblauer** [comment]
I would probably do something along these lines:

```js
// test.js
const t = require('tap');
const mock = require('mock-require');
const uuid = require('uuid/v4');

mock('uuid/v4', () => process.env.MOCK_UUID || uuid());

let subject = require('./subject');

t.test('should return uuid', t => {
  process.env.MOCK_UUID = '12345';

  const result = subject();

  t.equal(result, '12345');
  t.end();
});
```

You can jump through a bunch of hoops to make it work like you're doing now, but the method above is going to be much cleaner and definitely the path of least resistance.

**Comment by sarneeh** [comment]
@boblauer Hmm this brings some ideas to me 😄 

But let me first understand the issue first: is the problem caused by how Node caches `require` invocations? Considering the test from my very first message - the test doesn't pass because the process goes like this?

1. I'm requiring `./subject` at the start
2. The `./subject` require requires `./dep`, which requires `uuid/v4`, which is getting cached
3. I'm mocking `uuid/v4` and then resetting the cache of `./subject` but it doesn't do anything, as the cache is on `./dep` and it will still use the original version

Is this correct? 😄

**Comment by boblauer** [comment]
Correct. Every time something is required for the first time, it is cached, and all subsequent require statements will pull from the cache.

**Comment by sarneeh** [comment]
@boblauer One more question: why are you not decaching the mocked module everytime someone mocks it? It seems like it would be a lot easier and cleanier way to use it.

Then everytime someone does `mock('module', () => {})` will be sure that this particular module is not being cached anywhere and the next `require` will surely return the mocked object.

What were the reasons that you went with this library the way you went?

**Comment by boblauer** [comment]
I am decaching the mocked module every time someone mocks it. The problem is that if some other file already has a reference to the module, updating the cache won't matter. Here's a simple example:

```js
// five.js
module.exports = 5;

// lib.js
const five = require('./five');

// test.js
const lib = require('./lib');

mock('./five', 10);
```

When this code runs, `lib` will require `five`. At this point, updating the cache won't matter because `lib` is no longer referencing the cache. It's essentially the same as this:

```js
const five = 5;
const myValue = five;

five = 10;
```

Setting `five = 10` is not going to update `myValue`. This is essentially what's happening in your code above, which is why we need to re-require everything.

**Comment by sarneeh** [comment]
@boblauer I understand the issue. I just didn't explain properly what I've meant. So what I've really meant was that it would be cool to decache also all the children of the mocked module. This would prevent the need to `reRequire` all of the modules beneath (like in my example). [Decache](https://github.com/dwyl/decache) works like this.
